subsys/bluetooth/services/ans.c

```diff
@@ -0,0 +1,514 @@
+
+	uint8_t cat_in = (uint8_t)category;
+
+	/* Nothing to do if notify is disabled */
+	if (!new_alert_notify) {
+		return 0;
+	}
+
+	/* Special case: send all categories */
+	if (cat_in == BT_ANS_SEND_ALL_CATEGORY) {
+		for (int i = 0; i < BT_ANS_CAT_NUM; i++) {
+			if ((alert_sup_cat_bit_mask & BIT(i)) && (alert_cat_enabled_map & BIT(i))) {
+				rc = notify_new_alert_category(i);
+				if (rc < 0) {
+					return rc;
+				}
+			}
+		}
+		return 0;
+	}
+
+	/* Otherwise send just the requested category (if enabled) */
+	if (alert_cat_enabled_map & BIT(cat_in)) {
+		return notify_new_alert_category(cat_in);
+	}
+
+	return 0;
+}
+
+/* Helper to notify a single unread alert category */
+static int notify_unread_alert_category(uint8_t cat)
+{
+	int rc;
+	int ret;
+
+	ret = k_mutex_lock(&unread_mutex, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("Unable to lock mutex (err: %d)", ret);
+		return -EAGAIN;
+	}
+
+	rc = bt_gatt_notify_uuid(NULL, BT_UUID_GATT_UALRTS, ans_svc.attrs, &unread_alerts[cat],
+				 sizeof(struct unread_alert_status));
+
+	ret = k_mutex_unlock(&unread_mutex);
+	if (ret != 0) {
+		LOG_ERR("Unable to unlock mutex (err: %d)", ret);
+		return ret;
+	}
+
+	/* If the client is not connected, that is fine */
+	if (rc != 0 && rc != -ENOTCONN) {
+		LOG_ERR("Error notifying Unread Alert category %d rc: %d", cat, rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+/* Transmit Unread Alert */
+static int transmit_unread_alert(enum bt_ans_category category)
+{
+	int rc;
+
+	uint8_t cat_in = (uint8_t)category;
+
+	/* Nothing to do if notify is disabled */
+	if (!unread_alert_notify) {
+		return 0;
+	}
+
+	/* Special case: send all categories */
+	if (cat_in == BT_ANS_SEND_ALL_CATEGORY) {
+		for (int i = 0; i < BT_ANS_CAT_NUM; i++) {
+			if ((unread_sup_cat_bit_mask & BIT(i)) &&
+			    (unread_cat_enabled_map & BIT(i))) {
+				rc = notify_unread_alert_category(i);
+				if (rc < 0) {
+					return rc;
+				}
+			}
+		}
+		return 0;
+	}
+
+	/* Otherwise send just the requested category (if enabled) */
+	if (unread_cat_enabled_map & BIT(cat_in)) {
+		return notify_unread_alert_category(cat_in);
+	}
+
+	return 0;
+}
+
+int bt_ans_notify_new_alert(enum bt_ans_category category, uint8_t num_new, const char *text,
+			    int text_len)
+{
+	int ret;
+
+	uint8_t cat_in = (uint8_t)category;
+
+	/* Check if the category is supported */
+	if ((alert_sup_cat_bit_mask & (1U << cat_in)) == 0) {
+		LOG_WRN("Category %d unsupported", cat_in);
+		return -EINVAL;
+	}
+
+	/* Update the saved value */
+	ret = k_mutex_lock(&new_alert_mutex, K_FOREVER);
+	if (ret != 0) {
+		LOG_ERR("Unable to lock mutex (err: %d)", ret);
+		return ret;
+	}
+	new_alerts[cat_in].num_new_alerts = num_new;
+
+	int size = MIN(BT_ANS_MAX_TEXT_STR_SIZE - 1, text_len);
+
+	strncpy(new_alerts[cat_in].text_string, text, size);
+	new_alerts[cat_in].text_string[size] = '\0';
+	ret = k_mutex_unlock(&new_alert_mutex);
+	if (ret != 0) {
+		LOG_ERR("Unable to unlock mutex (err: %d)", ret);
+		return ret;
+	}
```

Comment:
Consider replacing this with

```suggestion
	__ASSERT(ret == 0, "Unable to unlock mutex (err: %d)", ret);
```

as if this fails to return, there is a fatal error in the code